File: packages/android/app/src/test/java/app/keyweb/sharing/BackupFileTest.kt
package app.keyweb.sharing

import app.keyweb.vault.HlcParts
import app.keyweb.vault.VaultOp
import app.keyweb.vault.VaultState
import app.keyweb.vault.applyOps
import app.keyweb.vault.datasetOf
import app.keyweb.vault.emptyVault
import app.keyweb.vault.encodeHlc
import app.keyweb.vault.field
import com.keyneom.synckit.sharing.ProtectedSharingIdentityV1
import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * A locked backup file, for the day Google is the thing that is down.
 *
 * The format is shared with the web, and that is most of what there is to
 * test: a file saved in a browser has to open on this phone and the other way
 * round, or the one copy somebody kept away from Google only works on whichever
 * device made it. So this opens a file the web really saved, and writes one of
 * its own for the web suite to open back.
 */
class BackupFileTest {

    private val json = Json { ignoreUnknownKeys = true }
    private val fixtures = File("../../../fixtures")

    @Serializable
    private data class LockFixture(
        val secretBase64: String,
        val keyId: String,
        val record: ProtectedSharingIdentityV1,
    )

    @Serializable
    private data class FileFixture(
        val note: String = "",
        val secretBase64: String,
        val keyId: String,
        val file: String,
    )

    private fun lock(): LockFixture =
        json.decodeFromString(LockFixture.serializer(), File(fixtures, "sharing-recovery-lock-v1.json").readText())

    private fun decode(base64: String) = java.util.Base64.getDecoder().decode(base64)

    private fun at(n: Long) = encodeHlc(HlcParts(1_700_000_000_000 + n, 0, "phone"))

    private fun vault(): VaultState = applyOps(
        emptyVault(),
        listOf(
            VaultOp.KeyringPut("k", at(0), "personal", "Just mine"),
            VaultOp.KeyringBind("b", at(1), "personal", "keyweb-abc"),
            VaultOp.ItemPut(
                "i",
                at(2),
                "bank",
                "personal",
                mapOf("title" to "Credit Union", "password" to "made-up-for-the-fixture"),
            ),
        ),
    )

    private fun saved(): Pair<String, ByteArray> {
        val lock = lock()
        val code = decode(lock.secretBase64)
        return sealBackupFile(vault(), unlockRecoveryLock(lock.record, code), lock.record) to code
    }

    @Test
    fun `opens with the printed code alone, and gives back the passwords`() {
        val (text, code) = saved()
        val opened = openBackupFile(text, code)
        assertEquals("made-up-for-the-fixture", opened.state.items.getValue("bank").field("password"))
        assertEquals(lock().keyId, opened.identity.publicKey.keyId)
    }

    @Test
    fun `opens a file the web saved`() {
        val fixture = json.decodeFromString(
            FileFixture.serializer(),
            File(fixtures, "backup-file-web-v1.json").readText(),
        )
        val opened = openBackupFile(fixture.file, decode(fixture.secretBase64))
        assertEquals("Credit Union", opened.state.items.getValue("bank").field("title"))
        assertEquals("Just mine", opened.state.keyrings.getValue("personal").name.value)
        assertEquals(fixture.keyId, opened.identity.publicKey.keyId)
    }

    @Test
    fun `hands back keyrings unbound`() {
        val (text, code) = saved()
        assertNull(datasetOf(openBackupFile(text, code).state.keyrings["personal"]))
    }

    @Test
    fun `says plainly when the code is wrong`() {
        val (text, _) = saved()
        assertFailsWith<WrongBackupCode> { openBackupFile(text, ByteArray(20)) }
    }

    @Test
    fun `says plainly when it is not a backup file at all`() {
        assertFailsWith<NotABackupFile> { openBackupFile("{\"items\":{}}", ByteArray(20)) }
    }

    @Test
    fun `never keeps the passwords readable in the file`() {
        val (text, _) = saved()
        assertTrue("made-up-for-the-fixture" !in text)
    }

    @Test
    fun `puts a keyring's passwords in the document of the file it lives in`() {
        val (text, code) = saved()
        val plan = restorePlan(vault(), openBackupFile(text, code).state)
        assertEquals(setOf("bank"), plan.getValue("keyweb-abc").items.keys)
        assertTrue("bank" !in plan.getValue("").items)
    }

    @Test
    fun `puts everything in the vault on a phone with nothing yet`() {
        val (text, code) = saved()
        val plan = restorePlan(emptyVault(), openBackupFile(text, code).state)
        assertEquals(setOf(""), plan.keys)
        assertTrue("bank" in plan.getValue("").items)
    }

    /** The file as the phone saves it, for the web suite to open back. */
    @Test
    fun `emits a file the phone saved, for the web suite`() {
        val lock = lock()
        val (text, _) = saved()
        File(fixtures, "backup-file-android-v1.json").writeText(
            Json { prettyPrint = true }.encodeToString(
                FileFixture.serializer(),
                FileFixture(
                    note = "Written by Android's sealBackupFile. Do not edit.",
                    secretBase64 = lock.secretBase64,
                    keyId = lock.keyId,
                    file = text,
                ),
            ) + "\n",
        )
    }
}

File: packages/web/tests/backup-file.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import type {
  ProtectedSharingIdentityStore,
  ProtectedSharingIdentityV1,
} from "@keyneom/sync-kit/sharing/web-passkey";
import { applyOps, datasetOf, emptyVault, encodeHlc, itemField } from "@keyweb/vault-core";
import { KEYWEB_RECOVERY_APP_ID, unlockRecoveryIdentity } from "../src/vault/sharing/identity";
import {
  openBackupFile,
  openBackupFileAsYou,
  restoreInto,
  sealBackupFile,
  WrongBackupCode,
} from "../src/vault/backupFile";

/**
 * An encrypted backup you keep yourself, opened with only the printed code.
 *
 * The case this exists for is Google being unreachable — an outage, a locked
 * account — with every device gone. The keyring files and the recovery lock in
 * Drive are no use then, so the file carries both halves: the vault sealed to
 * your identity, and the lock the code opens. These tests use the lock the
 * phone really writes (`sharing-recovery-lock-v1.json`), so they exercise the
 * cross-platform path rather than one the browser made for itself.
 */
const fixture = JSON.parse(
  readFileSync(new URL("../../../fixtures/sharing-recovery-lock-v1.json", import.meta.url), "utf8"),
) as { secretBase64: string; keyId: string; record: ProtectedSharingIdentityV1 };
const code = Uint8Array.from(Buffer.from(fixture.secretBase64, "base64"));

const appData: ProtectedSharingIdentityStore = {
  load: async (appId) => (appId === KEYWEB_RECOVERY_APP_ID ? fixture.record : null),
  save: async () => undefined,
  delete: async () => undefined,
};

function at(n: number) {
  return encodeHlc({ wall: 1_700_000_000_000 + n, counter: 0, node: "phone" });
}

/** A vault as it really is now: a keyring bound to its own Drive file. */
function vault() {
  return applyOps(emptyVault(), [
    { kind: "keyring.put", opId: "k", ts: at(0), keyringId: "personal", name: "Just mine" },
    { kind: "keyring.bind", opId: "b", ts: at(1), keyringId: "personal", datasetId: "keyweb-abc" },
    {
      kind: "item.put",
      opId: "i",
      ts: at(2),
      itemId: "bank",
      keyringId: "personal",
      fields: { title: "Credit Union", password: "the-one-that-matters" },
    },
  ]);
}

async function saved(): Promise<string> {
  const identity = (await unlockRecoveryIdentity(appData, code))!;
  const file = await sealBackupFile(vault(), identity, fixture.record);
  // As it would sit on a USB stick: text, not an object in memory.
  return JSON.stringify(file);
}

describe("an encrypted backup file", () => {
  it("opens with the printed code alone, and gives back the passwords", async () => {
    const restored = await openBackupFile(await saved(), code);
    expect(itemField(restored.items["bank"]!, "password")).toBe("the-one-that-matters");
    expect(restored.keyrings["personal"]!.name.value).toBe("Just mine");
  });

  it("hands back keyrings unbound, so they open with no Drive to reach", async () => {
    const restored = await openBackupFile(await saved(), code);
    expect(datasetOf(restored.keyrings["personal"])).toBeNull();
  });

  it("opens nothing with the wrong code", async () => {
    await expect(openBackupFile(await saved(), new Uint8Array(20).fill(9))).rejects.toThrow();
  });

  it("has nothing readable in it", async () => {
    const text = await saved();
    expect(text).not.toContain("the-one-that-matters");
    expect(text).not.toContain("Credit Union");
    expect(text).not.toContain("Just mine");
  });

  it("refuses a file that is not a Keyweb backup", async () => {
    await expect(openBackupFile('{"format":"something-else"}', code)).rejects.toThrow(
      "That isn't a Keyweb backup file.",
    );
  });

  it("refuses a lock that belongs to someone else", async () => {
    const identity = (await unlockRecoveryIdentity(appData, code))!;
    const stranger = {
      ...fixture.record,
      publicKey: { ...fixture.record.publicKey, keyId: "someone-else" },
    };
    await expect(sealBackupFile(vault(), identity, stranger)).rejects.toThrow(
      "different identity",
    );
  });
});

describe("restoring a backup file into this browser", () => {
  it("says plainly when the code is wrong", async () => {
    await expect(openBackupFileAsYou(await saved(), new Uint8Array(20))).rejects.toBeInstanceOf(
      WrongBackupCode,
    );
  });

  it("gives back you, and the lock the code opens", async () => {
    const opened = await openBackupFileAsYou(await saved(), code);
    expect(opened.identity.publicKey.keyId).toBe(fixture.keyId);
    expect(opened.recoveryLock.publicKey.keyId).toBe(fixture.keyId);
  });

  /** Two documents, the way a browser keeps a keyring that has its own file. */
  function memory(vaultDoc: ReturnType<typeof vault>) {
    const docs = new Map<string, ReturnType<typeof vault>>([["", vaultDoc]]);
    return {
      docs,
      readState: async (id = "") => docs.get(id) ?? emptyVault(),
      applyRemote: async (incoming: ReturnType<typeof vault>, id = "") => {
        docs.set(id, incoming);
        return incoming;
      },
    };
  }

  it("puts a keyring's passwords in its own file's document", async () => {
    const restored = await openBackupFile(await saved(), code);
    const store = memory(vault());
    await restoreInto(store, restored);
    expect(itemField(store.docs.get("keyweb-abc")!.items["bank"]!, "password")).toBe(
      "the-one-that-matters",
    );
    expect(store.docs.get("")!.items["bank"]).toBeUndefined();
  });

  it("puts everything in the vault on a browser with nothing yet", async () => {
    const restored = await openBackupFile(await saved(), code);
    const store = memory(emptyVault());
    await restoreInto(store, restored);
    expect(store.docs.get("")!.items["bank"]).toBeDefined();
  });
});

/**
 * The other direction: a file the phone saved, opened in a browser.
 *
 * Written by Android's `BackupFileTest`. The format is shared so that the one
 * copy somebody kept away from Google opens wherever they happen to be.
 */
describe("a backup file the phone saved", () => {
  const phone = JSON.parse(
    readFileSync(new URL("../../../fixtures/backup-file-android-v1.json", import.meta.url), "utf8"),
  ) as { secretBase64: string; keyId: string; file: string };

  it("opens here with the same code", async () => {
    const opened = await openBackupFileAsYou(
      phone.file,
      Uint8Array.from(Buffer.from(phone.secretBase64, "base64")),
    );
    expect(itemField(opened.state.items["bank"]!, "title")).toBe("Credit Union");
    expect(opened.state.keyrings["personal"]!.name.value).toBe("Just mine");
    expect(opened.identity.publicKey.keyId).toBe(phone.keyId);
  });
});
